jgrapht-core/src/main/java/org/jgrapht/generate/PlantedPartitionGraphGenerator.java

```diff
@@ -0,0 +1,310 @@
+    }
+
+    /**
+     * Construct a new PlantedPartitionGraphGenerator.
+     *
+     * @param l number of groups
+     * @param k number of nodes in each group
+     * @param p probability of connecting vertices within a group
+     * @param q probability of connecting vertices between groups
+     * @param rng random number generator
+     * @param selfLoops true if the graph allows self loops
+     * @throws IllegalArgumentException if number of groups is negative
+     * @throws IllegalArgumentException if number of nodes in each group is negative
+     * @throws IllegalArgumentException if p is not in [0,1]
+     * @throws IllegalArgumentException if q is not in [0,1]
+     */
+    public PlantedPartitionGraphGenerator(int l, int k, double p, double q, Random rng, boolean selfLoops)
+    {
+        if (l < 0) {
+            throw new IllegalArgumentException("number of groups must be non-negative");
+        }
+        if (k < 0) {
+            throw new IllegalArgumentException("number of nodes in each group must be non-negative");
+        }
+        if (p < 0 || p > 1) {
+            throw new IllegalArgumentException("invalid probability p");
+        }
+        if (q < 0 || q > 1) {
+            throw new IllegalArgumentException("invalid probability q");
+        }
+        this.l = l;
+        this.k = k;
+        this.p = p;
+        this.q = q;
+        this.rng = rng;
+        this.selfLoops = selfLoops;
+
+        this.fired = false;
+    }
+
+    /**
+     * Generate an $l$-planted partition graph.
+     *
+     * Note that the method can be called only once.
+     * Must instantiate another PlantedPartitionGraphGenerator object
+     * in order to generate another $l$-planted partition graph.
+     *
+     * @param target target graph
+     * @param resultMap result map
+     * @throws IllegalArgumentException if target is directed
+     * @throws IllegalArgumentException if self loops are requested but target does not allow them
+     * @throws IllegalStateException if generateGraph() is called more than once
+     */
+    @Override
+    public void generateGraph(Graph<V, E> target, Map<String, V> resultMap)
+    {
+        if (fired) {
+            throw new IllegalStateException("generateGraph() can be only called once");
+        }
+        this.fired = true;
+
+        // instantiate community structure
+        communities = new ArrayList<>(this.l);
+        for (int i = 0; i < this.l; i++) {
+            communities.add(new LinkedHashSet<>(this.k));
+        }
+
+        // empty graph case
+        if (this.l == 0 || this.k == 0) {
+            return;
+        }
+
+        // number of nodes
+        int n = this.k * this.l;
+        // integer to vertices
+        List<V> vertices = new ArrayList<>(n);
+        for (int i = 0; i < n; i++) {
+            V vertex = target.addVertex();
+            vertices.add(vertex);
+
+            // populate community structure
+            int lv = i / this.k;  // group of node v
```

Comment:
Can you double check this statement? This is an integer division: is this what you want? Instead of say 
`Math.ceil(1.0*i/k)`?

Reply:
Yep, integer division is what I want since I need the floor of the real division.